### DISCUSSION TITLE
[mini.ai] Trim extra whitespace between words?

### DISCUSSION THREAD

**Discussion by pkazmier** [nvim-mini/mini.nvim#1042] [Q&A] [answered]
If you had a sentence such as this with your cursor at the carat, how would you delete all but one space between the words?
```
Hello,         World
          ^
```
In emacs, I would regularly use `<A-Space>`, which deletes all spaces but one. I find it very convenient and have always missed this functionality in vim—likely because I don't know the equivalent way of doing so. And, I still am not sure of the "right" way to do this, so I built my first text object:

```lua
require("mini.ai").setup({
  custom_textobjects = {
    S = { { "^()()[ \t]*()()[^ \t]", "[^ \t]()[ \t]()[ \t]*()()[^ \t]" } },
  },
vim.keymap.set({"i", "n"}, "<A-Space>", "<Cmd>normal diS", "Just one space")
```

With the text object, I have both "a" and "i" variants. The "a" selects all spaces, and "i" selects all but one. And, then the mapping is just my shortcut to emulate the emacs behavior I've missed ever since moving to vim. So, is this overkill for my original task at hand? Thanks!

**Comment by echasnovski** [comment, marked as answer]
> So, is this overkill for my original task at hand? Thanks!

I would say so, yes. I tend yo use `ciw<Space><Esc>`, as `iw` works on whitespace too. So if you're fine with changing mode (twice) in a mapping, I'd go with `vim.keymap.set({"i", "n"}, "<A-Space>", "<Cmd>normal ciw <CR>", { desc = "Just one space" })`. If not, then custom function which executes `:normal diw` and then adds single space at cursor seems more understandable to me.

**Reply by pkazmier** [reply]
Thanks! I did not realize that `iw` worked on spaces. I guess it was overkill, but at least I learned how to write my own text objects with mini. The documentation was excellent and easy to understand!
